### PR TITLE
Job and DaemonSet documentation.

### PR DESCRIPTION
**Pull request by erictune**
Not ready for final review, but putting this out there so people know I am working on it.

### CODE REVIEW

File: docs/admin/daemon.yaml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+apiVersion: v1
```
Comment (erictune) [inline docs/admin/daemon.yaml]: @mikedanese @AnanyaKumar did either of you have examples planned for DaemonSet?

**Comment by erictune** [comment]
Ugh.  Somehow update-generated-docs.sh presubmit deleted all my documentation text.

**Comment by lavalamp** [comment]
@erictune if you ever committed it, you should be able to get it back, even if you rebased it away.

**Comment by davidopp** [comment]
@erictune Is this meant to replace #14079 or do you see them as complementary?

**Comment by davidopp** [comment]
Sorry I meant #13368

**Comment by pmorie** [comment]
@kubernetes/rh-cluster-infra

**Comment by erictune** [comment]
@mikedanese @soltysh @AnanyaKumar  this is ready for you to take a look at it

**Comment by davidopp** [comment]
Note that we're also going to be submitting Ananya's original design doc (#14326) but your doc makes sense as the user-consumable documentation.

**Comment by erictune** [comment]
@davidopp Sorry for slow response.  Yes, this was meant to user-focused and task-oriented -- a complement to the design doc.

File: docs/user-guide/jobs.md [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+
+Only a [`RestartPolicy`](pod-states.md) equal to `Never` or `OnFailure` are allowed.
+
+### Pod Selector
+
+The `.spec.selector` field is a pod selector.  It works the same as the `.spec.selector` of
+a [ReplicationController](replication-controller.md).
+
+If specified, the `.spec.template.metadata.labels` must be equal to the `.spec.selector`.  Config not following
+this rule will be rejected by the API.  If unspecified, it will be defaulted to the latter.
```
Comment (soltysh) [inline docs/user-guide/jobs.md]: I think the 'latter' should rather be the former in here, but it'll be best, to replace it with straight information that it takes value from `.spec.template.metadata.labels`.

File: docs/user-guide/jobs.md [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+[here](configuring-containers.md), and [here](working-with-resources.md).
+
+A Job also needs a [`.spec` section](../devel/api-conventions.md#spec-and-status).
+
+### Pod Template
+
+The `.spec.template` is the only required field of the `.spec`.
+
+The `.spec.template` is a [pod template](replication-controller.md#pod-template).
+ It has exactly the same schema as a [pod](pods.md), except
```
Comment (soltysh) [inline docs/user-guide/jobs.md]: Unnecessary space at the beginning.

File: docs/user-guide/jobs.md [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+this rule will be rejected by the API.  If unspecified, it will be defaulted to the latter.
+
+Also you should not normally create any pods whose labels match this selector, either directly,
+via another Job, or via another controller such as ReplicationController.  Otherwise, the Job will
+think that those pods were created by it.  Kubernetes will not stop you from doing this.
+
+### Multiple Completions
+
+By default, a Job is complete when one Pod runs to successful completion.  You can also specify that
+this needs to happen multiple times by specifying .spec.completions` with a value greater than 1.
```
Comment (soltysh) [inline docs/user-guide/jobs.md]: Missing apostrophe before `.spec.completions` brakes formatting.

Comment (soltysh) [inline docs/user-guide/jobs.md]: I wonder, because our current [validation](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/experimental/validation/validation.go#L286-L291) checks just for negative values, should we fix that to align with this statement?

File: docs/user-guide/jobs.md [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+### Parallelism
+
+You can suggest how many pods should run concurrently by setting `.spec.parallelism` to the number
+of pods you would like to have running concurrently.  This number is a suggestion. The number
+running concurrently may be lower or higher for a variety of reasons.  For example, it may be lower
+if the number of remaining completions is less, or as the controller is ramping up, or if it is
+throttling the job due to excessive failures.  It may be higher for example if a pod is gracefully
+shutdown, and the replacement starts early.
+
+If you do not specify `.spec.parallelism`, it is up to the Jon controller to chose a value.
```
Comment (soltysh) [inline docs/user-guide/jobs.md]: s/Jon/Job

Comment (soltysh) [inline docs/user-guide/jobs.md]: I'd state clearly we default it to 2.

File: docs/user-guide/jobs.md [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+
+## Handling Pod and Container Failures
+
+A Container in a Pod may fail for a number of reasons, such as because the process in it exited with
+a non-zero exit code, or the Container was killed for exceeding a memory limit, etc.  If this
+happens, and the `.spec.template.containers[].restartPolicy = "OnFailure"`, then the Pod stays
+on the node, but the Container is re-run.  Therefore, your program needs to handle the the case when it is
+restarted locally, or else specify `.spec.template.containers[].restartPolicy = "Never"`.
```
Comment (soltysh) [inline docs/user-guide/jobs.md]: Here again, I'm raising the issue of RestartPolicy, since my initial implementation differs from what we have right now, iow. now the RestartPolicy does not matter, we will always restart pods, which is not what the [API docs](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/experimental/types.go#L437-L439) says.

File: docs/user-guide/jobs.md [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+See [pods-states](pod-states.md) for more information on `restartPolicy`.
+
+An entire Pod can fail can also fail, for a number of reasons, such as when the pod is kicked off
```
Comment (soltysh) [inline docs/user-guide/jobs.md]: An entire Pod ~~can fail~~ can also fail,

File: docs/user-guide/jobs.md [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+starts a new Pod.  Therefore, your program needs to handle the case when it is restarted in a new
+pod.  In particular, it needs to handle temporary files, locks, incomplete output and the like
+caused by previous runs.
+
+Note that even if you specify `.spec.parallelism = 1` and `.spec.completions = 1` and
+`.spec.template.containers[].restartPolicy = "Never"`, the same program may
+sometimes be started twice.
+
+If you do specify `.spec.parallelism` and `.spec.completions` both greater than 1, then there may be
+multiple pods running at once.  Therefore, you pods must also be tolerant of concurrency.
```
Comment (soltysh) [inline docs/user-guide/jobs.md]: s/you/your

File: docs/user-guide/jobs.md [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+
+## Alternatives to Job
+
+### Bare Pods
+
+When a the node that a pod is running on reboots or fails, the pod is terminated
```
Comment (soltysh) [inline docs/user-guide/jobs.md]: When a ~~the~~ node...

File: docs/user-guide/jobs.md [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+requires only a single pod.
+
+### Replication Controller
+
+Jobs are complementary to [Replication Controllers](replication-controller.md).
+A Replication Controller manages pods which are not expected to terminate (e.g. web servers), and a Job
+manages pods that are expected to terminate (e.g. batch jobs).
+
+As discussed in [life of a pod](pod-states.md), `Job` is *only* appropriate for pods with
+`RestartPolicy` equal to o `OnFailure` or `Never`.  (Note: If `RestartPolicy` is not set, the default value is `Always`.)
```
Comment (soltysh) [inline docs/user-guide/jobs.md]: equal to ~~o~~ `OnFailure`

File: docs/user-guide/jobs.md [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+
+## Caveats
+
+Job is part of the experimental API group, so it is not subject to the same compatibility
+guarantees as objects in the main API.  It may not be enabled.  Enable by setting
+`--runtime-config=experimental/v1` on the apiserver.
+
+In general, pods which are created do not disappear until someone destroys them.  This might be a
+human or a `Jobs`, or another controller.  The only exception to this rule is that pods with a
```
Comment (soltysh) [inline docs/user-guide/jobs.md]: s/Jobs/Job

File: docs/user-guide/jobs.md [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+`PodPhase` of `Succeeded` or `Failed` for more than some duration (determined by the master) will
+expire and be automatically reaped.  But that automatic reaping interferes with completion counting.
```
Comment (soltysh) [inline docs/user-guide/jobs.md]: I'm not sure what you mean with the last statement. Maybe that should be:
"... be automatically reaped by the automatic reaping interfaces with completion counting." ?

**Comment by soltysh** [comment]
You're also missing running `hack/update-generated-docs.sh`, I recommend running `make verify` -  very handy :wink:.

File: docs/user-guide/jobs.md (hunk 2 above)
Comment (erictune) [inline docs/user-guide/jobs.md]: fixed in next push

File: docs/user-guide/jobs.md (hunk 3 above)
Comment (erictune) [inline docs/user-guide/jobs.md]: fixed

File: docs/user-guide/jobs.md (hunk 4 above)
Comment (erictune) [inline docs/user-guide/jobs.md]: I think an explicit value of 1 is allowed.  Is that what you are asking?

File: docs/user-guide/jobs.md (hunk 5 above)
Comment (erictune) [inline docs/user-guide/jobs.md]: Made changes.

File: docs/user-guide/jobs.md (hunk 6 above)
Comment (erictune) [inline docs/user-guide/jobs.md]: Even with the changes by Mike, RestartPolicy still affects whether certain failures turn into just container restarts, or complete Pod restarts.

File: docs/user-guide/jobs.md (hunk 7 above)
Comment (erictune) [inline docs/user-guide/jobs.md]: fixed

File: docs/user-guide/jobs.md (hunk 5 above)
Comment (erictune) [inline docs/user-guide/jobs.md]: Discussing options for handling the defaulting of parallelism with @bgrant0607.  Leaving somewhat vague for now.

File: docs/user-guide/jobs.md (hunk 8 above)
Comment (erictune) [inline docs/user-guide/jobs.md]: fixed

File: docs/user-guide/jobs.md (hunk 9 above)
Comment (erictune) [inline docs/user-guide/jobs.md]: fix

File: docs/user-guide/jobs.md (hunk 10 above)
Comment (erictune) [inline docs/user-guide/jobs.md]: fix

File: docs/user-guide/jobs.md (hunk 11 above)
Comment (erictune) [inline docs/user-guide/jobs.md]: fix.

File: docs/user-guide/jobs.md (hunk 12 above)
Comment (erictune) [inline docs/user-guide/jobs.md]: Removed this section.  I don't think we have any automatic api pod reaping ATM.

**Comment by soltysh** [comment]
Apparently shippable doesn't like me a lot in this PR :wink:

File: docs/user-guide/jobs.md (hunk 4 above)
Comment (soltysh) [inline docs/user-guide/jobs.md]: I'm reading your statement as: valid values are 1 and above, whereas our validation allows 0 and above, thus the question is should we leave validation as is for completions or change it to be 1 and above, because it doesn't make sense to have 0 completions does it?

File: docs/user-guide/jobs.md (hunk 6 above)
Comment (soltysh) [inline docs/user-guide/jobs.md]: At which level is this achieved? I'd like to read about it, if you could point me to docs, that will be greatly appreciated.

File: pkg/controller/job/controller.go [hunk 13]
```diff
@@ -341,7 +341,7 @@ func (jm *JobController) syncJob(key string) error {
 		job.Status.Unsuccessful = unsuccessful
 
 		if err := jm.updateHandler(&job); err != nil {
-			glog.Errorf("Failed to update job %v, requeuing.  Error: %v", job.Name, err)
+			glog.Errorf("Failed to update job %v, requeuing", job.Name)
```
Comment (soltysh) [inline pkg/controller/job/controller.go]: Is this change intentional, you've just introduced this with other PR.

**Comment by soltysh** [comment]
Job LGTM, except for the controller being modified by mistake(?).

**Comment by mikedanese** [comment]
ref #1518

File: docs/user-guide/jobs.md (hunk 6 above)
Comment (erictune) [inline docs/user-guide/jobs.md]: docs/user-guide/pod-states.md describes restartPolicy.  Is that what you are asking?

File: pkg/controller/job/controller.go (hunk 13 above)
Comment (erictune) [inline pkg/controller/job/controller.go]: thanks, this was an accident.  Fixed in next push.

File: docs/user-guide/jobs.md (hunk 6 above)
Comment (soltysh) [inline docs/user-guide/jobs.md]: Yup, thanks.

**Comment by soltysh** [comment]
Now you have my 100% LGTM :smile:

**Comment by erictune** [comment]
@k8s-oncall this is only docs and test code and comments.  You may merge despite free if you like.

File: docs/admin/daemons.md [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+create pods on nodes which match that [node
+selector](../user-guide/node-selection/README.md).
+
+If you do not specify a `.spec.template.spec.nodeSelector`, then the DaemonSet controller will
+create pods on all nodes.
+
+## How Daemon Pods are Scheduled
+
+Normally, the machine that a pod runs on is selected by the Kubernetes scheduler.  However, pods
+create by the Daemon controller have the machine already selected (`.spec.nodeName` is specified
```
Comment (derekwaynecarr) [inline docs/admin/daemons.md]: pods created by

File: docs/admin/daemons.md [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
+node failure or disruptive node maintenance, such as a kernel upgrade. For this reason, you should
+use a Daemon Set rather than creating individual pods.
+
+### Static Pods
+
+It is possible to create pods by writing a file to a certain directory watched by Kubelet.  These
+are called [static pods](static-pods.md).
+Unlike DaemonSet, static pods cannot be managed with kubectl
+or other Kubernetes API clients.  Static pods do not depend on the apiserver, making them useful
+in cluster bootstrapping cases.  Also, static pods may be deprecated in the future.
```
Comment (derekwaynecarr) [inline docs/admin/daemons.md]: I need to verify this, but if your cluster rejects the ability to run privileged containers in the api-server, you could have a configuration where the kubelet enables the ability to run privileged containers.  As a result, you could have a configuration where the use of static pods could be enabled to support running a privileged container, but the daemonset would be rejected from running a privileged container.

Comment (erictune) [inline docs/admin/daemons.md]: Yes, I think you could have that situation.

**Comment by bgrant0607** [comment]
Needs rebase

File: docs/admin/daemon.yaml [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+kind: DaemonSet
+metadata:
+  name: prometheus-node-exporter
+spec:
+  template:
+    metadata:
+      name: prometheus-node-exporter
+      labels:
+        name: prometheus-node-exporter
```
Comment (bgrant0607) [inline docs/admin/daemon.yaml]: Please don't use "name" as a label key.
https://github.com/kubernetes/kubernetes/blob/master/docs/user-guide/config-best-practices.md

How about something like `daemon: monitoring`? If I were a cluster admin, I'd probably use the same label key for all my node agents / daemons, but not `name`. `name` is too generic (it could overlap with non-daemon pods) and confusing for users (due to similarity with metadata.name).

File: pkg/apis/experimental/v1alpha1/types.go [hunk 17]
```diff
@@ -400,17 +403,21 @@ type JobSpec struct {
...
 	Completions *int `json:"completions,omitempty"`
 
 	// Selector is a label query over pods that should match the pod count.
+	// More info: http://releases.k8s.io/HEAD/docs/user-guide/labels.md#label-selectors
 	Selector map[string]string `json:"selector,omitempty"`
 
 	// Template is the object that describes the pod that will be created when
 	// executing a job.
+	// More info: http://releases.k8s.io/HEAD/docs/user-guide/jobs.md
 	Template *v1.PodTemplateSpec `json:"template"`
```
Comment (bgrant0607) [inline pkg/apis/experimental/v1alpha1/types.go]: The field should be a pointer if and only if the field is optional.

File: docs/admin/daemon.yaml (hunk 16 above)
Comment (erictune) [inline docs/admin/daemon.yaml]: fixed.

File: docs/admin/daemons.md (hunk 14 above)
Comment (erictune) [inline docs/admin/daemons.md]: fixed

File: pkg/apis/experimental/v1alpha1/types.go (hunk 17 above)
Comment (erictune) [inline pkg/apis/experimental/v1alpha1/types.go]: Noted, but not getting fixed in this PR.

**Comment by erictune** [comment]
Fixed to use `apiVersion: experimental/v1alpha1`

File: pkg/apis/experimental/v1alpha1/types.go (hunk 17 above)
Comment (soltysh) [inline pkg/apis/experimental/v1alpha1/types.go]: Fixed in https://github.com/kubernetes/kubernetes/pull/14575
